*A Course in Cryptography* by Heiko Knospe, American Mathematical Society, Pure and Applied Undergraduate Texts 40

## Code examples of Chapter 1 - Fundamentals

This SageMath notebook by Heiko Knospe is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.

Download SageMath from http://www.sagemath.org/download.

SageMath can work with large integers.

In [1]:
2^128

340282366920938463463374607431768211456

Computing a residue class.

In [2]:
mod(-892342322327,11)

6

Algebraic normal form of a Boolean Function.

In [3]:
from sage.crypto.boolean_function import BooleanFunction
B=BooleanFunction([1,1,0,1]) # Boolean function defined by f(00)=1, f(01)=1, f(10)=0, f(11)=1
B.algebraic_normal_form()

x0*x1 + x1 + 1

Print the binomials $\binom{15}{n}$.

In [4]:
for n in range(0,16):
	print(binomial(15,n))

1
15
105
455
1365
3003
5005
6435
6435
5005
3003
1365
455
105
15
1


Compute the probability that no collision occurs if $k$ samples are drawn from a set of $n$ elements.

In [5]:
def birthday(k,n):
    p=1
    for i in range(1,k):
        p=p*(1-i/n)
    print(RR(p))

The probability of a birthday collision for $23$ people is around $50\%$. 

In [6]:
birthday(23,365)

0.492702765676015


Find a collision in the output of the recursion $x_{i}=f(x_{i-1})$ where $f(x)=x^2+26 \mod 107$ and $x_0=1$. Use Floyd's cycle finding algorithm.

In [7]:
def f(x):
    return(x*x+26)    
x=mod(1,107); y=mod(1,107)
x=f(x)
y=f(f(y))
k=1
while x!=y:    
    x=f(x)
    y=f(f(y)) 
    k=k+1
print("k ={:3}, x = {}".format(k,x))

k =  9, x = 39


Check the collision and print $x_1, x_2, \dots$. We see that $x_9=39$ and $x_{18}=39$.

In [8]:
x=mod(1,107)
for i in range(18):
    x=f(x)
    print("{}".format(x)),

27
6
62
18
29
11
40
21
39
49
73
5
51
59
83
67
21
39


Exercise 9.

In [9]:
from sage.crypto.boolean_function import BooleanFunction
BooleanFunction([1,1,1,0,0,0,1,1]).algebraic_normal_form()

x0*x1*x2 + x0*x1 + x1*x2 + x2 + 1

Exercise 15.

In [10]:
def f(x):
    return(x*x+1)    
x=mod(1,56807); y=mod(1,56807)
x=f(x)
y=f(f(y))
k=1
while x!=y:    
    x=f(x)
    y=f(f(y)) 
    k=k+1
print("k ={:4}, x = {}".format(k,x))

k = 184, x = 22604


$184$ is a period. Now we compute the least period.

In [11]:
x0=mod(22604,56807)
x=f(x0)
k=1
while x!=x0:
    x=f(x)
    k=k+1
print("period ={:3}".format(k))    

period = 46
